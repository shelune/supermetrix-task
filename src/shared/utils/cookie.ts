export function setCookie(name: string, val: string, duration: number) {
  const date = new Date();
  const value = val;

  // duration in minutes
  date.setTime(date.getTime() + duration * 60 * 1000);

  document.cookie = `${name}=${value}; expires=${date.toUTCString()}; path=/`;
}

export function getCookie(name: string): string | undefined {
  const value = `; ${document.cookie}`;
  const parts = value.split(`; ${name}=`);

  if (parts.length === 2) {
    return parts.pop()?.split(";").shift();
  }
  return undefined;
}

export function deleteCookie(name: string) {
  const date = new Date();

  // Set it expire in -1 days
  date.setTime(date.getTime() + -1 * 24 * 60 * 60 * 1000);

  document.cookie = `${name}=; expires=${date.toUTCString()}; path=/`;
}
