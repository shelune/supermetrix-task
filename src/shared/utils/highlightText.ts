import { createElement, DetailedReactHTMLElement } from "react";

export function escapeRegExp(string: string): string {
  // $& means the whole matched string
  return string.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

export function highlightText(
  string: string,
  highlight: string,
  highlightClass = "",
): (
  | DetailedReactHTMLElement<
      {
        className: string;
      },
      HTMLElement
    >
  | string
)[] {
  if (!string || typeof string !== "string") {
    return ["-"];
  }
  if (!highlight || highlight.length === 0) return [string];
  const reg = new RegExp(escapeRegExp(highlight), "gi");
  const res = [];
  let lastIndex = 0;
  let result;
  // eslint-disable-next-line no-cond-assign
  while ((result = reg.exec(string))) {
    res.push(string.substring(lastIndex, result.index));
    res.push(
      createElement(
        "span",
        { className: highlightClass, key: `${highlight}-${lastIndex}` },
        [result[0]],
      ),
    );
    lastIndex = result.index + result[0].length;
  }
  res.push(string.substring(lastIndex));
  return res;
}
