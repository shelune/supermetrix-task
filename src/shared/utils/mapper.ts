import type { Post, ApiPost } from "../api";

export const postMapper = ({
  id,
  from_name,
  from_id,
  message,
  type,
  created_time,
}: ApiPost): Post => ({
  id,
  authorName: from_name,
  authorId: from_id,
  message,
  type,
  createdTime: new Date(created_time),
});
