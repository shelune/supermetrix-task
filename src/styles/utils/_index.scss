@import "_variables.scss";
@import "_typography.scss";
@import "_utilities.scss";
