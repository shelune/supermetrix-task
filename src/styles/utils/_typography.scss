@mixin text-title-md {
  font-size: 22px;
  line-height: 30px;
  letter-spacing: -0.15px;
}

@mixin text-body {
  font-size: 16px;
  line-height: 20px;
}

@mixin text-subtitle-md {
  font-size: 13px;
  line-height: 16px;
}

@mixin text-subtitle-sm {
  font-size: 11px;
  line-height: 12px;
}

@mixin text-label {
  font-size: 11px;
  line-height: 12px;
  text-transform: uppercase;
}
