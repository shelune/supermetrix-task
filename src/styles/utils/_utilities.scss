$breakpoint-mobile: 768px;
$breakpoint-desktop: 1024px;

@mixin mobile {
  @media (max-width: #{$breakpoint-mobile}) {
    @content;
  }
}

@mixin tablet {
  @media (max-width: #{$breakpoint-desktop}) and (min-width: #{$breakpoint-mobile}) {
    @content;
  }
}

@mixin desktop {
  @media (min-width: #{$breakpoint-desktop}) {
    @content;
  }
}

@mixin icon-base {
  display: inline-block;
  cursor: pointer;
  @content;

  svg {
    width: 100%;
    height: 100%;
  }
}

@mixin icon-md {
  @include icon-base;
  width: 24px;
  height: 24px;
}

@mixin shadow-light {
  box-shadow: 0px 0.5px 0px #eeeeee;
}

@mixin rounded-4 {
  border-radius: 4px;
}

@mixin rounded-8 {
  border-radius: 8px;
}

@mixin rounded-full {
  border-radius: 9999px;
}
