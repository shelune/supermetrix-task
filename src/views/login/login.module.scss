@import "../../styles/utils/_index.scss";

.loginPage {
  display: flex;
  align-items: center;
  justify-content: center;
  height: 100vh;
}

.container {
  width: 436px;
  margin: 0 auto;
  text-align: center;
  padding: 0 24px;

  @include mobile() {
    width: 100%;
  }
}

.formControl {
  width: 100%;
  margin-bottom: 20px;
  position: relative;
}

.label {
  @include text-body;
  color: $gray;
  position: absolute;
  left: 16px;
  top: 22.5px;

  &.isEditing {
    @include text-label;
    top: 14.5px;
  }
}

.textInput {
  @include rounded-4;
  @include text-body;
  width: 100%;
  max-width: 100%;
  padding: 27.5px 16px 19.5px;
  border: 1px solid $gray;
}

.showPassword {
  cursor: pointer;
  width: 16px;
  height: 16px;
  position: absolute;
  right: 16px;
  top: 22.5px;
  opacity: 0.5;

  &.isActive {
    opacity: 1;
  }
}

.formFunctions {
  margin-top: 40px;
}

.submitButton {
  @include text-body;
  @include rounded-4;
  background-color: $orange;
  color: $white;
  outline: none;
  border: none;
  height: 64px;
  line-height: 64px;
  width: 100%;
}

.formError {
  color: $red-error;
}
