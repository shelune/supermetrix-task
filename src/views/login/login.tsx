import classNames from "classnames";
import axios from "axios";
import { useNavigate } from "react-router-dom";
import React, { FC, useCallback, useState } from "react";
import type { ApiErrorResponse, ApiRegisterResponse } from "../../shared/api";
import { setCookie } from "../../shared/utils/cookie";

import css from "./login.module.scss";

type Props = {
  setToken: (token: string | null) => void;
};

export const LoginView: FC<Props> = ({ setToken }) => {
  const [username, setUsername] = useState("");
  const [email, setEmail] = useState("");
  const navigate = useNavigate();
  const [isSubmitting, setSubmitting] = useState(false);
  const [error, setError] = useState("");

  const login = useCallback(async () => {
    setSubmitting(true);
    try {
      const response = await axios.post<ApiRegisterResponse>(
        "https://api.supermetrics.com/assignment/register",
        {
          client_id: "ju16a6m81mhid5ue1z3v2g0uh",
          email,
          name: username,
        },
      );
      if (response.data.data.sl_token) {
        setCookie("sl_token", response.data.data.sl_token, 60);
        setToken(response.data.data.sl_token);
        navigate("/posts");
      }
    } catch (err) {
      if (axios.isAxiosError(err) && err.response?.data) {
        const errorResponse = err.response.data as ApiErrorResponse;
        setError(errorResponse.error.message);
      } else {
        // eslint-disable-next-line no-console
        console.log("unknown error when registering", err);
        setError("unknown error when registering");
      }
    }

    setSubmitting(false);
  }, [email, navigate, setToken, username]);

  return (
    <div className={css.loginPage}>
      <div className={css.container}>
        <form
          className={css.form}
          onSubmit={(e) => {
            e.preventDefault();
            // eslint-disable-next-line @typescript-eslint/no-floating-promises
            login();
          }}
        >
          <div className={css.formControl}>
            <label
              htmlFor="username"
              className={classNames(css.label, {
                [css.isEditing]: !!username,
              })}
            >
              Name
            </label>
            <input
              className={css.textInput}
              type="text"
              value={username}
              id="username"
              name="username"
              onChange={(e) => {
                setUsername(e.target.value);
              }}
            />
          </div>
          <div className={css.formControl}>
            <label
              htmlFor="email"
              className={classNames(css.label, {
                [css.isEditing]: !!email,
              })}
            >
              Email
            </label>
            <input
              className={css.textInput}
              type="email"
              value={email}
              id="email"
              name="email"
              onChange={(e) => {
                setEmail(e.target.value);
              }}
            />
          </div>
          {error && (
            <div data-testid="login-error" className={css.formError}>
              {error}
            </div>
          )}
          <div className={css.formFunctions}>
            <button
              type="submit"
              className={css.submitButton}
              data-testid="login-submit"
            >
              {isSubmitting ? "Submitting" : "Submit"}
            </button>
          </div>
        </form>
      </div>
    </div>
  );
};
