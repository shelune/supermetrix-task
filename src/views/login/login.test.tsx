import axios from "axios";
import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";
import React, { ReactNode } from "react";
import { act, fireEvent, render } from "@testing-library/react";
import { LoginView } from "./login";
import { setCookie } from "../../shared/utils/cookie";

jest.mock("axios");
jest.mock("../../shared/utils/cookie", () => ({
  setCookie: jest.fn(),
}));

const mockedUseNavigate = jest.fn();

// eslint-disable-next-line @typescript-eslint/no-unsafe-return
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUseNavigate,
}));

const mockedAxios = axios as jest.Mocked<typeof axios>;

function renderWithRouter(children: ReactNode) {
  const history = createMemoryHistory();
  return render(
    <Router location={history.location} navigator={history}>
      {children}
    </Router>,
  );
}

describe("LoginView", () => {
  it("should render", async () => {
    // eslint-disable-next-line @typescript-eslint/await-thenable
    const { container } = await renderWithRouter(
      <LoginView setToken={jest.fn()} />,
    );
    expect(container).toMatchSnapshot();
  });

  it("should render input accordingly if it has value", () => {
    const { getByLabelText, getByText } = render(
      <LoginView setToken={jest.fn()} />,
    );
    act(() => {
      fireEvent.change(getByLabelText("Email"), {
        target: {
          value: "hello",
        },
      });
    });
    expect(getByText("Email")).toMatchSnapshot();
  });

  it("should render error if login fails", async () => {
    mockedAxios.post.mockReturnValueOnce(Promise.reject(new Error("Huh")));

    // eslint-disable-next-line @typescript-eslint/await-thenable
    const { getByTestId } = await renderWithRouter(
      <LoginView setToken={jest.fn()} />,
    );

    await act(async () => {
      // eslint-disable-next-line @typescript-eslint/await-thenable
      await getByTestId("login-submit").click();
    });
    expect(getByTestId("login-error")).toBeInTheDocument();
  });

  it("should behave correctly if login succeeds", async () => {
    const mockSetToken = jest.fn();
    // eslint-disable-next-line @typescript-eslint/await-thenable
    const { getByTestId } = await renderWithRouter(
      <LoginView setToken={mockSetToken} />,
    );
    mockedAxios.post.mockReturnValueOnce(
      Promise.resolve({
        data: {
          data: {
            sl_token: "token",
          },
        },
      }),
    );
    await act(async () => {
      // eslint-disable-next-line @typescript-eslint/await-thenable
      await getByTestId("login-submit").click();
    });
    expect(mockSetToken).toHaveBeenCalledWith("token");
    expect(mockedUseNavigate).toHaveBeenCalledWith("/posts");
    expect(setCookie).toHaveBeenCalled();
  });
});
