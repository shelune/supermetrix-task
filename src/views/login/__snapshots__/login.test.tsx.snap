// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginView should render 1`] = `
<div>
  <div
    class="loginPage"
  >
    <div
      class="container"
    >
      <form
        class="form"
      >
        <div
          class="formControl"
        >
          <label
            class="label"
            for="username"
          >
            Name
          </label>
          <input
            class="textInput"
            id="username"
            name="username"
            type="text"
            value=""
          />
        </div>
        <div
          class="formControl"
        >
          <label
            class="label"
            for="email"
          >
            Email
          </label>
          <input
            class="textInput"
            id="email"
            name="email"
            type="email"
            value=""
          />
        </div>
        <div
          class="formFunctions"
        >
          <button
            class="submitButton"
            data-testid="login-submit"
            type="submit"
          >
            Submit
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`LoginView should render input accordingly if it has value 1`] = `
<label
  class="label isEditing"
  for="email"
>
  Email
</label>
`;
